lib/manifold/api/workspace.rb

```diff
@@ -0,0 +1,106 @@
+# frozen_string_literal: true
+
+module Manifold
+  module API
+    # Encapsulates a single manifold.
+    class Workspace
+      attr_reader :name, :template_path
+
+      DEFAULT_TEMPLATE_PATH = File.expand_path(
+        "../templates/workspace_template.yml", __dir__
+      ).freeze
+
+      def initialize(name, template_path: DEFAULT_TEMPLATE_PATH, logger: Logger.new($stdout))
+        self.name = name
+        self.template_path = template_path
+        @logger = logger
+        @vector_service = Services::VectorService.new(logger)
+      end
+
+      def self.from_directory(directory, logger: Logger.new($stdout))
+        new(directory.basename.to_s, logger: logger)
+      end
+
+      def add
+        [tables_directory, routines_directory].each(&:mkpath)
+        FileUtils.cp(template_path, manifold_path)
+      end
+
+      def generate
+        return unless manifold_exists? && any_vectors?
+
+        generate_dimensions
+        @logger.info("Generated BigQuery dimensions table schema for workspace '#{name}'.")
+      end
+
+      def tables_directory
+        directory.join("tables")
+      end
+
+      def routines_directory
+        directory.join("routines")
+      end
+
+      def manifold_file
+        return nil unless manifold_exists?
+
+        File.new(manifold_path)
+      end
+
+      def manifold_exists?
+        manifold_path.file?
+      end
+
+      def manifold_path
+        directory.join("manifold.yml")
+      end
+
+      private
+
+      def directory
+        Pathname.pwd.join("workspaces", name)
+      end
+
+      def manifold_yaml
+        @manifold_yaml ||= YAML.safe_load_file(manifold_path)
+      end
+
+      def generate_dimensions
+        dimensions_path.write(dimensions_schema_json.concat("\n"))
+      end
+
+      def dimensions_schema
+        [
+          { "type" => "STRING", "name" => "id", "mode" => "REQUIRED" },
+          { "type" => "RECORD", "name" => "dimensions", "mode" => "REQUIRED",
+            "fields" => dimensions_fields }
+        ]
+      end
+
+      def dimensions_fields
+        vectors.reduce([]) do |list, vector|
+          @logger.info("Loading vector schema for '#{vector}'.")
+          list << @vector_service.load_vector_schema(vector)
+        end
```

Comment:
A few observations here. First, I find it's best to avoid using mutating methods or operators (eg `<<`) unless you have a specific reason for using one. It might seem trivial, but as a guiding principle it results in much more maintainable code. The original code was fine. If you need to change the order then just flip it around.

Second, in the original code you splatted the result of `load_vector_schema`, which meant that `nil`s would be silently ignored. `load_vector_schema` can return `nil`, so is it really valid to include them in your `dimensions_fields`? Is it even valid to have a vector without a schema? Perhaps you should be raising an exception?

Finally, stepping back a little, why not use `filter_map` or `map` here? `reduce` is usually for reducing a list to a non-list.
```suggestion
        vectors.filter_map do |vector|
          @logger.info("Loading vector schema for '#{vector}'.")
          @vector_service.load_vector_schema(vector)
        end
```